### ISSUE TITLE
`unnecessary_semicolon` FP on expression statements with attributes

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#15426]
### Summary

`unnecessary_semicolon` suggests removing a semicolon that is necessary to convince the compiler to permit an attribute on an expression statement.

### Lint Name

unnecessary_semicolon

### Reproducer

I tried this code:

```rust
#![warn(clippy::unnecessary_semicolon)]

pub fn example(x: i32) {
    #[rustfmt::skip]
    match x {
        0b00 => {}  0b01 => {}
        0b11 => {}  _    => {}
    };
}
```

I saw this happen:

```
warning: unnecessary semicolon
 --> src/lib.rs:8:6
  |
8 |     };
  |      ^ help: remove
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_semicolon
```

I expected to see this happen:
No lint, or an alternate suggestion, because removing the semicolon is a syntax error.

### Version

```text
rustc 1.89.0 (29483883e 2025-08-04)
binary: rustc
commit-hash: 29483883eed69d5fb4db01964cdf2af4d86e9cb2
commit-date: 2025-08-04
host: aarch64-apple-darwin
release: 1.89.0
LLVM version: 20.1.7


Also reproduces on 1.91.0-nightly (2025-08-05 ec7c02612527d185c379)
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by ada4a** [comment]
(somewhat) related: https://github.com/rust-lang/rust-clippy/issues/15388
